sycl/source/detail/graph_impl.hpp

```diff
@@ -308,11 +465,98 @@ class graph_impl {
   /// @param Queue In-order queue to register \p Node for.
   /// @param Node Last node that was added to this graph from \p Queue.
   void setLastInorderNode(std::shared_ptr<sycl::detail::queue_impl> Queue,
-                             std::shared_ptr<node_impl> Node) {
+                          std::shared_ptr<node_impl> Node) {
     std::weak_ptr<sycl::detail::queue_impl> QueueWeakPtr(Queue);
     MInorderQueueMap[QueueWeakPtr] = Node;
   }
 
+  /// Prints the contents of the graph to a text file in DOT format
+  /// @param GraphName is a string appended to the output file name
+  void printGraphAsDot(const char *GraphName) {
+    /// Vector of nodes visited during the graph printing
+    std::vector<node_impl *> MVisitedNodes;
+    std::string GraphNameStr(GraphName);
+    std::string FileName = "graph_" + GraphNameStr + ".dot";
+
+    std::fstream Stream(FileName, std::ios::out);
+    Stream << "strict digraph {" << std::endl;
+
+    for (std::shared_ptr<node_impl> Node : MRoots)
+      Node->printDotRecursive(Stream, MVisitedNodes);
+
+    Stream << "}" << std::endl;
+
+    for (auto Node : MVisitedNodes)
+      Node->MVisited = false;
+  }
+
+  /// Checks if the graph_impl of Graph has a similar structure to
+  /// the graph_impl of the caller.
+  /// Graphs are considered similar if they have same numbers of nodes
+  /// of the same type with similar predecessor and successor nodes (number and
+  /// type). Two nodes are considered similar if they have the same
+  /// command-group type. For command-groups of type "kernel", the "signature"
+  /// of the kernel is also compared (i.e. the name of the command-group).
+  /// @param Graph if reference to the graph to compare with.
+  /// @param DebugPrint if set to true print additional debug information about
+  /// the spotted graph differences
+  /// @return true if the two graphs are similar, false otherwise
+  bool hasSimilarStructure(std::shared_ptr<detail::graph_impl> Graph,
+                           bool DebugPrint = false) const {
+    if (this == Graph.get())
+      return true;
+
+    if (MContext != Graph->MContext) {
+      if (DebugPrint)
+        std::cout << "MContext are not the same." << std::endl;
```

Comment:
I'd prefer we didn't print anything to `std::cout` it's generally bad practice for libraries to print to standard out. Can we throw an exception instead with this information in the exception.